src/core/thread/mle.cpp

```diff
@@ -1725,51 +1720,82 @@ ThreadError Mle::HandleDataResponse(const Message &aMessage, const Ip6::MessageI
+        // send MLE Data Request
+        if ((timestamp == NULL || timestamp->Compare(activeTimestamp) != 0) &&
+            (Tlv::GetOffset(aMessage, Tlv::kActiveDataset, activeDatasetOffset) != kThreadError_None))
         {
-            aMessage.Read(offset, sizeof(tlv), &tlv);
-            mNetif.GetActiveDataset().Set(activeTimestamp, aMessage, offset + sizeof(tlv), tlv.GetLength());
-        }
-        else
-        {
-            dataRequest = true;
+            ExitNow(dataRequest = true);
         }
     }
+    else
+    {
+        activeTimestamp.SetLength(0);
+    }
 
     // Pending Timestamp
     if (Tlv::GetTlv(aMessage, Tlv::kPendingTimestamp, sizeof(pendingTimestamp), pendingTimestamp) == kThreadError_None)
     {
+        const MeshCoP::Timestamp *timestamp;
+
         VerifyOrExit(pendingTimestamp.IsValid(), error = kThreadError_Parse);
+        timestamp = mNetif.GetPendingDataset().GetNetwork().GetTimestamp();
 
-        // Pending Dataset
-        if (Tlv::GetOffset(aMessage, Tlv::kPendingDataset, offset) == kThreadError_None)
+        // if received timestamp does not match the local value and message does not contain the dataset,
+        // send MLE Data Request
+        if ((timestamp == NULL || timestamp->Compare(pendingTimestamp) != 0) &&
+            (Tlv::GetOffset(aMessage, Tlv::kPendingDataset, pendingDatasetOffset) != kThreadError_None))
         {
-            aMessage.Read(offset, sizeof(tlv), &tlv);
-            mNetif.GetPendingDataset().Set(activeTimestamp, aMessage, offset + sizeof(tlv), tlv.GetLength());
-        }
-        else
-        {
-            dataRequest = true;
+            ExitNow(dataRequest = true);
         }
     }
+    else
+    {
+        pendingTimestamp.SetLength(0);
+    }
 
     // Network Data
     mNetworkData.SetNetworkData(leaderData.GetDataVersion(), leaderData.GetStableDataVersion(),
                                 (mDeviceMode & ModeTlv::kModeFullNetworkData) == 0,
                                 networkData.GetNetworkData(), networkData.GetLength());
 
+    // Active Dataset
+    if (activeTimestamp.GetLength() > 0)
+    {
+        aMessage.Read(activeDatasetOffset, sizeof(tlv), &tlv);
+        mNetif.GetActiveDataset().Set(activeTimestamp, aMessage, activeDatasetOffset + sizeof(tlv), tlv.GetLength());
+    }
+    else
+    {
+        mNetif.GetActiveDataset().GetNetwork().Clear();
+    }
+
+    // Pending Dataset
+    if (pendingTimestamp.GetLength() > 0)
+    {
+        aMessage.Read(pendingDatasetOffset, sizeof(tlv), &tlv);
+        mNetif.GetPendingDataset().Set(pendingTimestamp, aMessage, pendingDatasetOffset + sizeof(tlv), tlv.GetLength());
+    }
+    else
+    {
+        mNetif.GetPendingDataset().GetNetwork().Clear();
+    }
+
+    mRetrieveNewNetworkData = false;
+
 exit:
     (void)aMessageInfo;
 
-    if (diff > 0 && dataRequest)
+    if (dataRequest)
     {
-        uint8_t tlvs[] = {Tlv::kNetworkData};
+        static const uint8_t tlvs[] = {Tlv::kNetworkData};
```

Comment:
I don't see any place in the spec that requires including the Leader Data TLV in a MLE Data Request.